File: content-solutions/graph-advanced/uf-reverse-cuts.java
import java.util.*;

class Solution {
    int[] parent, size;

    void init(int n) {
        parent = new int[n];
        size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }

    int find(int x) {
        int root = x;
        while (parent[root] != root) root = parent[root];
        while (x != root) {
            int next = parent[x];
            parent[x] = root;
            x = next;
        }
        return root;
    }

    boolean union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return false;
        if (size[ra] < size[rb]) {
            int t = ra;
            ra = rb;
            rb = t;
        }
        parent[rb] = ra;
        size[ra] += size[rb];
        return true;
    }

    public int[] solution(int n, int[][] bridges, int[] cuts) {
        init(n);
        boolean[] cut = new boolean[bridges.length];
        for (int c : cuts) cut[c] = true;
        int groups = n;
        for (int i = 0; i < bridges.length; i++) {
            if (!cut[i] && union(bridges[i][0], bridges[i][1])) groups--;
        }
        int[] answer = new int[cuts.length];
        for (int k = cuts.length - 1; k >= 0; k--) {
            answer[k] = groups;
            int[] b = bridges[cuts[k]];
            if (union(b[0], b[1])) groups--;
        }
        return answer;
    }
}

File: content-solutions/graph-advanced/uf-reverse-cuts.js
class DSU {
  constructor(n) {
    this.parent = Array.from({ length: n }, (_, i) => i);
    this.size = new Array(n).fill(1);
  }
  find(x) {
    let root = x;
    while (this.parent[root] !== root) root = this.parent[root];
    while (x !== root) {
      const next = this.parent[x];
      this.parent[x] = root;
      x = next;
    }
    return root;
  }
  union(a, b) {
    let ra = this.find(a);
    let rb = this.find(b);
    if (ra === rb) return false;
    if (this.size[ra] < this.size[rb]) [ra, rb] = [rb, ra];
    this.parent[rb] = ra;
    this.size[ra] += this.size[rb];
    return true;
  }
}

function solution(n, bridges, cuts) {
  const dsu = new DSU(n);
  const cut = new Set(cuts);
  let groups = n;
  bridges.forEach(([a, b], i) => {
    if (!cut.has(i) && dsu.union(a, b)) groups--;
  });
  const answer = [];
  for (let k = cuts.length - 1; k >= 0; k--) {
    answer.push(groups);
    const [a, b] = bridges[cuts[k]];
    if (dsu.union(a, b)) groups--;
  }
  return answer.reverse();
}
